src/ElfUtils/ElfFile.cpp

```diff
@@ -197,34 +230,14 @@ ErrorMessageOr<ModuleSymbols> ElfFileImpl<ElfT>::LoadSymbols() {
   module_symbols.set_load_bias(load_bias);
   module_symbols.set_symbols_file_path(file_path_.string());
 
-  for (const llvm::object::ELFSymbolRef& symbol_ref : object_file_->symbols()) {
-    if ((symbol_ref.getFlags() & llvm::object::BasicSymbolRef::SF_Undefined) != 0) {
-      continue;
-    }
-    std::string name = symbol_ref.getName() ? symbol_ref.getName().get() : "";
-    std::string demangled_name = llvm::demangle(name);
-
-    // Unknown type - skip and generate a warning
-    if (!symbol_ref.getType()) {
-      LOG("WARNING: Type is not set for symbol \"%s\" in \"%s\", skipping.", name,
-          file_path_.string());
-      continue;
-    }
-
-    // Limit list of symbols to functions. Ignore sections and variables.
-    if (symbol_ref.getType().get() != llvm::object::SymbolRef::ST_Function) {
-      continue;
-    }
-
-    uint64_t symbol_address = symbol_ref.getValue();
-    SymbolInfo* symbol_info = module_symbols.add_symbol_infos();
-    symbol_info->set_name(name);
-    symbol_info->set_demangled_name(demangled_name);
-    symbol_info->set_address(symbol_address);
-    symbol_info->set_size(symbol_ref.getSize());
+  for (const llvm::object::ELFSymbolRef& symbol_ref : object_file_->getDynamicSymbolIterators()) {
```

Comment:
this only makes sense if .symtab is not present. The dynamic symbol set is always a subset of .symtab

Reply:
(It might make sense to split it into LoadSymbols, LoadSymbolsFromSymtab, LoadSymbolsFromDynsym depending on your use-case, with LoadSymbols trying to load fromSymtab and if it is not available falling back to dynsym)

Reply:
splitted into two methods as suggested